lib/internal/crypto/hash.js

```diff
@@ -140,7 +146,74 @@ Hmac.prototype.digest = function digest(outputEncoding) {
 Hmac.prototype._flush = Hash.prototype._flush;
 Hmac.prototype._transform = Hash.prototype._transform;
 
+class Mac extends LazyTransform {
+  constructor(mac, ...args) {
+    validateString(mac, 'mac');
+
+    let nid = EVP_PKEY_HMAC;
+    let alg, key, options;
+
+    switch (mac) {
+      case 'cmac':
+        nid = EVP_PKEY_CMAC;
+        // Fall through.
+
+      case 'hmac':
+        [alg, key, options] = args;
+        validateString(alg, 'alg');
+        break;
+
+      case 'poly1305':
+        [key, options] = args;
+        nid = EVP_PKEY_POLY1305;
+        break;
+
+      case 'siphash':
+        [key, options] = args;
+        nid = EVP_PKEY_SIPHASH;
+        break;
+
+      default:
+        throw new ERR_INVALID_OPT_VALUE('mac', mac);
+    }
+
+    key = prepareSecretKey(key);
+    super(options);
+
+    this[kHandle] = new _Mac(nid, toBuf(key), alg);
+  }
+
+  _transform(chunk, encoding, callback) {
+    this[kHandle].update(chunk, encoding);
```

Comment:
nit... likely would be worthwhile to assert here that chunk is the expected type. For most Node.js core cases that will always be true but there are some incorrect Transform implementations out in the wild that sometimes pass unexpected types.

Reply:
Yeah, I'm going to leave that out because we don't check that anywhere else either, AFAIK. That kind of change should be done in one big sweep.